### PR TITLE
Add selectable dates limitation feature

### LINKED ISSUE #197: Limit the range of selectable dates (min and/or max)
The developer should be able to limit the dates the user is allowed to select.

The most common use case for "min" is probably to limit the selection to future dates.

One use case for "max" is to limit the selection to the end of the current year or one year ahead.

If min > max, no days should be selectable.

Both min and max are inclusive, meaning the user is able to select the minimum or maximum date specified.

Draft code:
```html
<!-- Limit selection to "today – end of year" -->
<vaadin-date-picker min="2016-06-17" max="2016-12-31"></vaadin-date-picker>
```
Feel free to come up with better property/attribute names, if you see these confusing or problematic.

The format for the `min` and `max` properties follows the same rules as the `value` property.

If a value outside min or max is set programmatically (or via data binding), the field should be set to invalid state, and an error message should be shown below the field ("Date is not allowed"). The validation message should be added to the i18n object.

### PR DESCRIPTION
**Pull request by samiheikki**
Fixes #197

---
This change is [Reviewable](https://reviewable.io/reviews/vaadin/vaadin-date-picker/235)

### CODE REVIEW

**Comment by tehapo** [comment]
Reviewed 10 of 10 files at r1.
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---

*[vaadin-date-picker-behavior.html, line 292 at r1](https://reviewable.io:443/reviews/vaadin/vaadin-date-picker/235#-KRYLL6AJQHI1cDKs2-C:-KRYLL6AJQHI1cDKs2-D:b4tnak8) ([raw file](https://github.com/vaadin/vaadin-date-picker/blob/84280c6f5eafb6c9cb742592330219a032188269/vaadin-date-picker-behavior.html#L292)):*
> ```HTML
> 
>     _minChanged: function(value, oldValue) {
>       if (!value) {
> ```

Instead of repeating the whole `_valueChanged` functionality here and in `_maxChanged`, I would suggest some kind of helper function(s) that could be used by `_valueChanged`, `_minChanged` and `_maxChanged`.

---

*[vaadin-date-picker-overlay.html, line 660 at r1](https://reviewable.io:443/reviews/vaadin/vaadin-date-picker/235#-KRYOt7HpkrTX9oCzjZE:-KRYOt7HpkrTX9oCzjZF:b-teptlx) ([raw file](https://github.com/vaadin/vaadin-date-picker/blob/84280c6f5eafb6c9cb742592330219a032188269/vaadin-date-picker-overlay.html#L660)):*
> ```HTML
>               break;
>             case 'home':
>               if (this.minDate) {
> ```

How about simplifying this a bit? And the `end` case could be refactored similarily.

```
dateToFocus = new Date(focus.getFullYear(), focus.getMonth(), 1);
if (this.minDate && dateToFocus < this.minDate) {
  this.focusedDate = this.minDate;
} else {
  this.focusedDate = dateToFocus;
}
```

---

*[vaadin-month-calendar.html, line 82 at r1](https://reviewable.io:443/reviews/vaadin/vaadin-date-picker/235#-KRYPTnb6uWXvRw2YeAT:-KRYPTnb6uWXvRw2YeAU:b-xmueat) ([raw file](https://github.com/vaadin/vaadin-date-picker/blob/84280c6f5eafb6c9cb742592330219a032188269/vaadin-month-calendar.html#L82)):*
> ```HTML
> 
>       #monthGrid div[disabled] {
>         color: #9e9e9e;
> ```

I think using something like `opacity: 0.3` would be better than fixing a shade of grey here.

---

*[demo/index.html, line 61 at r1](https://reviewable.io:443/reviews/vaadin/vaadin-date-picker/235#-KRYQc4XQNEk6rVUH27i:-KRYQc4YTZ-gpHfUTkek:b-7ok57i) ([raw file](https://github.com/vaadin/vaadin-date-picker/blob/84280c6f5eafb6c9cb742592330219a032188269/demo/index.html#L61)):*
> ```HTML
>     </demo-snippet>
> 
>     <h3>Earliest and latest date</h3>
> ```

I think using the exact terms min and max in the title would be more clear here.

---

*[docs/vaadin-date-picker-value.adoc, line 40 at r1](https://reviewable.io:443/reviews/vaadin/vaadin-date-picker/235#-KRYTFpb3b5TYJ_xwWQX:-KRYTFpb3b5TYJ_xwWQY:b-p43b68) ([raw file](https://github.com/vaadin/vaadin-date-picker/blob/84280c6f5eafb6c9cb742592330219a032188269/docs/vaadin-date-picker-value.adoc#L40)):*
> ```AsciiDoc
> ----
> 
> == Limit the Selectable values
> ```

Shouldn't the word "values" be capitalized also ("Values")? 

---

*[test/basic.html, line 495 at r1](https://reviewable.io:443/reviews/vaadin/vaadin-date-picker/235#-KRYR3k_7GaDmb2er-0Z:-KRYR3k_7GaDmb2er-0_:bm85gk9) ([raw file](https://github.com/vaadin/vaadin-date-picker/blob/84280c6f5eafb6c9cb742592330219a032188269/test/basic.html#L495)):*
> ```HTML
>           datepicker.value = '2017-01-01';
>           expect(datepicker.invalid).to.be.equal(true);
>         });
> ```

How about also testing a case where a date is valid (inside the defined range)? And maybe also testing the corner cases of exact date of min or max?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-date-picker/235)*

**Comment by samiheikki** [comment]
Review status: 4 of 10 files reviewed at latest revision, 6 unresolved discussions.

---

*[vaadin-date-picker-behavior.html, line 292 at r1](https://reviewable.io:443/reviews/vaadin/vaadin-date-picker/235#-KRYLL6AJQHI1cDKs2-C:-KRY_YeAfY9S72UprMyn:b-896fix) ([raw file](https://github.com/vaadin/vaadin-date-picker/blob/84280c6f5eafb6c9cb742592330219a032188269/vaadin-date-picker-behavior.html#L292)):*
<details><summary><i>Previously, tehapo (Teemu Pöntelin) wrote…</i></summary>
> Instead of repeating the whole `_valueChanged` functionality here and in `_maxChanged`, I would suggest some kind of helper function(s) that could be used by `_valueChanged`, `_minChanged` and `_maxChanged`.
</details>
Done.

---

*[vaadin-date-picker-overlay.html, line 660 at r1](https://reviewable.io:443/reviews/vaadin/vaadin-date-picker/235#-KRYOt7HpkrTX9oCzjZE:-KRYYKaYy-h81qRtq7C9:b-896fix) ([raw file](https://github.com/vaadin/vaadin-date-picker/blob/84280c6f5eafb6c9cb742592330219a032188269/vaadin-date-picker-overlay.html#L660)):*
<details><summary><i>Previously, tehapo (Teemu Pöntelin) wrote…</i></summary>
> How about simplifying this a bit? And the `end` case could be refactored similarily.
> 
> ```
> dateToFocus = new Date(focus.getFullYear(), focus.getMonth(), 1);
> if (this.minDate && dateToFocus < this.minDate) {
>   this.focusedDate = this.minDate;
> } else {
>   this.focusedDate = dateToFocus;
> }
> ```
</details>
Done.

---

*[vaadin-month-calendar.html, line 82 at r1](https://reviewable.io:443/reviews/vaadin/vaadin-date-picker/235#-KRYPTnb6uWXvRw2YeAT:-KRYWWxwoemcNk_XsKBe:b-896fix) ([raw file](https://github.com/vaadin/vaadin-date-picker/blob/84280c6f5eafb6c9cb742592330219a032188269/vaadin-month-calendar.html#L82)):*
<details><summary><i>Previously, tehapo (Teemu Pöntelin) wrote…</i></summary>
> I think using something like `opacity: 0.3` would be better than fixing a shade of grey here.
</details>
Done.

---

*[demo/index.html, line 61 at r1](https://reviewable.io:443/reviews/vaadin/vaadin-date-picker/235#-KRYQc4XQNEk6rVUH27i:-KRYWYBHd5EC9PXARpwO:b-896fix) ([raw file](https://github.com/vaadin/vaadin-date-picker/blob/84280c6f5eafb6c9cb742592330219a032188269/demo/index.html#L61)):*
<details><summary><i>Previously, tehapo (Teemu Pöntelin) wrote…</i></summary>
> I think using the exact terms min and max in the title would be more clear here.
</details>
Done.

---

*[docs/vaadin-date-picker-value.adoc, line 40 at r1](https://reviewable.io:443/reviews/vaadin/vaadin-date-picker/235#-KRYTFpb3b5TYJ_xwWQX:-KRYWYsp2UMAC8Jdad4m:b-896fix) ([raw file](https://github.com/vaadin/vaadin-date-picker/blob/84280c6f5eafb6c9cb742592330219a032188269/docs/vaadin-date-picker-value.adoc#L40)):*
<details><summary><i>Previously, tehapo (Teemu Pöntelin) wrote…</i></summary>
> Shouldn't the word "values" be capitalized also ("Values")? 
</details>
Done.

---

*[test/basic.html, line 495 at r1](https://reviewable.io:443/reviews/vaadin/vaadin-date-picker/235#-KRYR3k_7GaDmb2er-0Z:-KRYXTrJ905AHGFDrFp5:b-896fix) ([raw file](https://github.com/vaadin/vaadin-date-picker/blob/84280c6f5eafb6c9cb742592330219a032188269/test/basic.html#L495)):*
<details><summary><i>Previously, tehapo (Teemu Pöntelin) wrote…</i></summary>
> How about also testing a case where a date is valid (inside the defined range)? And maybe also testing the corner cases of exact date of min or max?
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-date-picker/235)*

**Comment by tehapo** [comment]
:lgtm:

---

Reviewed 6 of 6 files at r2.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-date-picker/235#-:-KRbXGC96LfD1JHRqiTc:bnfp4nl)*

**Comment by tomivirkki** [comment]
*[vaadin-date-picker-behavior.html, line 451 at r2](https://reviewable.io:443/reviews/vaadin/vaadin-date-picker/235#-KRbXrO-waE6uFLlpUOB:-KRbXrO-waE6uFLlpUOC:bsnzbyh) ([raw file](https://github.com/vaadin/vaadin-date-picker/blob/826a4e43b960b31cd54ad0e2728b1021fd24383c/vaadin-date-picker-behavior.html#L451)):*
> ```HTML
> 
>     _dateAllowed: function(date, min, max) {
>       if (min && max) {
> ```

Similar checks occur in multiple (3) locations. It might be justifiable to have the `_dateAllowed` method in all layers separately but in that case I'd rather minimize the code repetition to a one-liner:
`return (!min || date >= min) && (!max || date <= max);`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-date-picker/235)*

**Comment by tehapo** [comment]
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[vaadin-date-picker-behavior.html, line 451 at r2](https://reviewable.io:443/reviews/vaadin/vaadin-date-picker/235#-KRbXrO-waE6uFLlpUOB:-KRbZG2dQSs10SDbwz09:bz6c886) ([raw file](https://github.com/vaadin/vaadin-date-picker/blob/826a4e43b960b31cd54ad0e2728b1021fd24383c/vaadin-date-picker-behavior.html#L451)):*
<details><summary><i>Previously, tomivirkki (Tomi Virkki) wrote…</i></summary>
> Similar checks occur in multiple (3) locations. It might be justifiable to have the `_dateAllowed` method in all layers separately but in that case I'd rather minimize the code repetition to a one-liner:
> `return (!min || date >= min) && (!max || date <= max);`
</details>
:+1:
Good point @tomivirkki. I'll take my :lgtm: back. :laughing: 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-date-picker/235)*

**Comment by tomivirkki** [comment]
Also noticed an issue with keyboard navigation: If the focus is initially on a disallowed location (like for example if you leave out the initial-position attribute from the demo or set the date-picker value programmatically) the navigation gets stuck. Could move the focus to allowed value in '_moveFocusByDays' and '_moveFocusByMonths' if the date is out of allowed range. Remember to add tests.

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-date-picker/235#-:-KRbc-_Yi488ZE61CgZ2:b-q1arp)*

**Comment by samiheikki** [comment]
Done.

---

Review status: 7 of 10 files reviewed at latest revision, 1 unresolved discussion.

---

*[vaadin-date-picker-behavior.html, line 451 at r2](https://reviewable.io:443/reviews/vaadin/vaadin-date-picker/235#-KRbXrO-waE6uFLlpUOB:-KRcObnyf-DLJYZE36wu:b-896fix) ([raw file](https://github.com/vaadin/vaadin-date-picker/blob/826a4e43b960b31cd54ad0e2728b1021fd24383c/vaadin-date-picker-behavior.html#L451)):*
<details><summary><i>Previously, tehapo (Teemu Pöntelin) wrote…</i></summary>
> :+1:
> Good point @tomivirkki. I'll take my :lgtm: back. :laughing: 
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-date-picker/235#-:-KRcO_yfjjm3ylg-CFqs:b-896fix)*

**Comment by tehapo** [comment]
Reviewed 3 of 3 files at r3.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[vaadin-date-picker-overlay.html, line 762 at r3](https://reviewable.io:443/reviews/vaadin/vaadin-date-picker/235#-KReMLY6Gh5gnf4Dk-R6:-KReMLY6Gh5gnf4Dk-R7:b-flpa55) ([raw file](https://github.com/vaadin/vaadin-date-picker/blob/491b01623ccd2af8f26fd8a8e7c62ea2e1477b8c/vaadin-date-picker-overlay.html#L762)):*
> ```HTML
>       },
> 
>       _moveFocusToEndOfMonth: function(focusedDate) {
> ```

This method is so similar to the `_moveFocusToStartOfMonth` that I think you could combine these to a single one that takes some extra arguments?

---

*[vaadin-date-picker-overlay.html, line 788 at r3](https://reviewable.io:443/reviews/vaadin/vaadin-date-picker/235#-KReKr7fJR6-gnoBWeyr:-KReKr7fJR6-gnoBWeys:b9osybk) ([raw file](https://github.com/vaadin/vaadin-date-picker/blob/491b01623ccd2af8f26fd8a8e7c62ea2e1477b8c/vaadin-date-picker-overlay.html#L788)):*
> ```HTML
>         var closestDate;
> 
>         dates.forEach(function(value) {
> ```

Hmm.. I'm not sure I like the way `forEach` is used here. You are essentially trying to _filter_ out undefined values and then _reduce_ the array to a single item which is the closest date? :wink: 

So basically this whole method could maybe look like:
```
_getClosestDate: function(date, dates) {
  return dates.filter(function(date) {
           return date !== undefined;
         })
         .reduce(function(closestDate, candidate) {
           var candidateDiff = Math.abs(date.getTime() - candidate.getTime());
           var closestDateDiff = Math.abs(closestDate.getTime() - date.getTime());
           return candidateDiff < closestDateDiff ? candidate : closestDate;
         });
}
```

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-date-picker/235)*

**Comment by samiheikki** [comment]
Review status: 9 of 10 files reviewed at latest revision, 3 unresolved discussions.

---

*[vaadin-date-picker-overlay.html, line 762 at r3](https://reviewable.io:443/reviews/vaadin/vaadin-date-picker/235#-KReMLY6Gh5gnf4Dk-R6:-KRgEz6yupD1sei8IEj5:b-896fix) ([raw file](https://github.com/vaadin/vaadin-date-picker/blob/491b01623ccd2af8f26fd8a8e7c62ea2e1477b8c/vaadin-date-picker-overlay.html#L762)):*
<details><summary><i>Previously, tehapo (Teemu Pöntelin) wrote…</i></summary>
> This method is so similar to the `_moveFocusToStartOfMonth` that I think you could combine these to a single one that takes some extra arguments?
</details>
Done.

---

*[vaadin-date-picker-overlay.html, line 788 at r3](https://reviewable.io:443/reviews/vaadin/vaadin-date-picker/235#-KReKr7fJR6-gnoBWeyr:-KRgILJjT8uN0VOdm6Gx:b-896fix) ([raw file](https://github.com/vaadin/vaadin-date-picker/blob/491b01623ccd2af8f26fd8a8e7c62ea2e1477b8c/vaadin-date-picker-overlay.html#L788)):*
<details><summary><i>Previously, tehapo (Teemu Pöntelin) wrote…</i></summary>
> Hmm.. I'm not sure I like the way `forEach` is used here. You are essentially trying to _filter_ out undefined values and then _reduce_ the array to a single item which is the closest date? :wink: 
> 
> So basically this whole method could maybe look like:
> ```
> _getClosestDate: function(date, dates) {
>   return dates.filter(function(date) {
>            return date !== undefined;
>          })
>          .reduce(function(closestDate, candidate) {
>            var candidateDiff = Math.abs(date.getTime() - candidate.getTime());
>            var closestDateDiff = Math.abs(closestDate.getTime() - date.getTime());
>            return candidateDiff < closestDateDiff ? candidate : closestDate;
>          });
> }
> ```
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-date-picker/235)*

**Comment by tehapo** [comment]
Originally the issue #197 also asked for a default error message to be displayed when an invalid date is set as the value. However we discussed with @jouni that it might be better to leave a default message out.

Still, it would be great to have an example in the documentation (`vaadin-date-picker-value.adoc`) on how to assign an error message. At least my initial idea would be to have an `invalid-changed` listener which then sets the `errorMessage` property of the `vaadin-date-picker`.

Could you still add this example to this PR? Thanks! :)

---

Reviewed 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-date-picker/235#-:-KRgp6EQ92_YejRcj9-4:b1ra45u)*

**Comment by tehapo** [comment]
Nevermind, discussed things further and decided to create a separate issue on improving our documentation related to validation and error handling. :lgtm: 

---

Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-date-picker/235#-:-KRgw1KeXMD4pqD7qbet:bo0jzlb)*
